File: src/main/java/com/dev/farmmanager/service/transaction/TransactionService.java
package com.dev.farmmanager.service.transaction;

import com.dev.farmmanager.domain.dto.transaction.TransactionPageDto;
import com.dev.farmmanager.domain.entity.Transaction;
import com.dev.farmmanager.domain.enumeration.TransactionStatus;
import com.dev.farmmanager.domain.enumeration.TransactionType;
import com.dev.farmmanager.domain.payload.transaction.TransactionPayload;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public interface TransactionService {

    TransactionPageDto findAll(Integer cropCycleId, TransactionType type, LocalDate date, TransactionStatus status);

    Optional<Transaction> getById(Integer id);

    List<Transaction> findExpensesByCropCycleId(Integer cropCycleId);

    List<Transaction> findIncomesByCropCycleId(Integer cropCycleId);

    Transaction create(Integer userId, TransactionPayload payload);

    Transaction update(Integer id, TransactionPayload payload);

    void delete(Integer id);
}

File: frontend/src/services/transaction.service.ts
import http from '@/api/http'
import type { TransactionDto, TransactionPage, TransactionPayload, TransactionStatus, TransactionType } from '@/types/transaction'

export interface TransactionFilters {
  crop_cycle_id?: number | null
  type?: TransactionType | null
  date?: string | null
  status?: TransactionStatus | null
}

export const transactionService = {
  findAll: (filters?: TransactionFilters) => {
    const params: Record<string, string> = {}
    if (filters?.crop_cycle_id != null) params.crop_cycle_id = String(filters.crop_cycle_id)
    if (filters?.type)   params.type   = filters.type
    if (filters?.date)   params.date   = filters.date
    if (filters?.status) params.status = filters.status
    return http.get<TransactionPage>('/farm/transaction', { params }).then((r) => r.data)
  },

  create: (payload: TransactionPayload) =>
    http.post('/farm/transaction', payload).then((r) => r.data),

  getById: (id: number) =>
    http.get<TransactionDto>(`/farm/transaction/${id}`).then((r) => r.data),

  update: (id: number, payload: TransactionPayload) =>
    http.put(`/farm/transaction/${id}`, payload).then((r) => r.data),

  delete: (id: number) =>
    http.delete(`/farm/transaction/${id}`),
}
